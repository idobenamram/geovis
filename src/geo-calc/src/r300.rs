use serde::{Deserialize, Serialize};
use std::f64::consts::PI;
/// taken from https://bivector.net/tools.html?p=3&q=0&r=0
// Written by a generator written by enki.
use std::fmt;
use std::ops::{Add, BitAnd, BitOr, BitXor, Index, IndexMut, Mul, Not, Sub};
use wasm_bindgen::prelude::*;

const basis: &'static [&'static str] = &["1", "e1", "e2", "e3", "e12", "e13", "e23", "e123"];
const basis_count: usize = basis.len();

#[wasm_bindgen]
#[derive(Clone, Copy, PartialEq, Debug, Serialize, Deserialize)]
pub struct R300 {
    mvec: [f64; basis_count],
}

impl R300 {
    pub const fn zero() -> Self {
        Self {
            mvec: [0.0; basis_count],
        }
    }

    pub const fn new(f: f64, idx: usize) -> Self {
        let mut ret = Self::zero();
        ret.mvec[idx] = f;
        ret
    }
}

#[wasm_bindgen]
impl R300 {

    pub fn vector(e1_val: f64, e2_val: f64, e3_val: f64) -> Self {
        let mut ret = Self::zero();
        ret.mvec[1] = e1_val;
        ret.mvec[2] = e2_val;
        ret.mvec[3] = e3_val;
        ret
    }

    pub fn bivector(e12_val: f64, e13_val: f64, e23_val: f64) -> Self {
        let mut ret = Self::zero();
        ret.mvec[4] = e12_val;
        ret.mvec[5] = e13_val;
        ret.mvec[6] = e23_val;
        ret
    }

    pub fn display(&self) -> String {
        let mut parts = Vec::new();
        
        for (i, prefix) in basis.iter().enumerate() {
            parts.push(format!("{}: {}", prefix, self.mvec[i]));
        }
        format!("R300({})", parts.join(", "))

    }

    #[wasm_bindgen(js_name = toJson)]
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap()
    }
}

// basis vectors are available as global constants.
const e1: R300 = R300::new(1.0, 1);
const e2: R300 = R300::new(1.0, 2);
const e3: R300 = R300::new(1.0, 3);
const e12: R300 = R300::new(1.0, 4);
const e13: R300 = R300::new(1.0, 5);
const e23: R300 = R300::new(1.0, 6);
const e123: R300 = R300::new(1.0, 7);

impl Index<usize> for R300 {
    type Output = f64;

    fn index<'a>(&'a self, index: usize) -> &'a Self::Output {
        &self.mvec[index]
    }
}

impl IndexMut<usize> for R300 {
    fn index_mut<'a>(&'a mut self, index: usize) -> &'a mut Self::Output {
        &mut self.mvec[index]
    }
}

impl fmt::Display for R300 {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut n = 0;
        let ret = self
            .mvec
            .iter()
            .enumerate()
            .filter_map(|(i, &coeff)| {
                if coeff > 0.00001 || coeff < -0.00001 {
                    n = 1;
                    Some(format!(
                        "{}{}",
                        format!("{:.*}", 7, coeff)
                            .trim_end_matches('0')
                            .trim_end_matches('.'),
                        if i > 0 { basis[i] } else { "" }
                    ))
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
            .join(" + ");
        if n == 0 {
            write!(f, "0")
        } else {
            write!(f, "{}", ret)
        }
    }
}

impl R300 {
    // Reverse
    // Reverse the order of the basis blades.
    pub fn Reverse(self: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0];
        res[1] = a[1];
        res[2] = a[2];
        res[3] = a[3];
        res[4] = -a[4];
        res[5] = -a[5];
        res[6] = -a[6];
        res[7] = -a[7];
        res
    }

    // Dual
    // Poincare duality operator.
    pub fn Dual(self: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = -a[7];
        res[1] = -a[6];
        res[2] = a[5];
        res[3] = -a[4];
        res[4] = a[3];
        res[5] = -a[2];
        res[6] = a[1];
        res[7] = a[0];
        res
    }

    // Conjugate
    // Clifford Conjugation
    pub fn Conjugate(self: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0];
        res[1] = -a[1];
        res[2] = -a[2];
        res[3] = -a[3];
        res[4] = -a[4];
        res[5] = -a[5];
        res[6] = -a[6];
        res[7] = a[7];
        res
    }

    // Involute
    // Main involution
    pub fn Involute(self: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0];
        res[1] = -a[1];
        res[2] = -a[2];
        res[3] = -a[3];
        res[4] = a[4];
        res[5] = a[5];
        res[6] = a[6];
        res[7] = -a[7];
        res
    }
    pub fn bracket(self, blade: u8) -> R300 {
        R300::new(self[blade as usize], blade as usize)
    }

    // taken from New Foundation for Classical Mechanics - David Hestenes - page 61 (Magnitude)
    pub fn magnitude_squared(self: Self) -> f64 {
        let scalar_part = (self * self.Reverse())[0];

        scalar_part
    }

    pub fn norm(self: Self) -> f64 {
        let scalar_part = (self * self.Conjugate())[0];

        scalar_part.abs().sqrt()
    }

    pub fn inorm(self: Self) -> f64 {
        self.Dual().norm()
    }

    pub fn normalized(self: Self) -> Self {
        self * (1.0 / self.norm())
    }

    pub fn dot(self: Self, b: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = b[0] * a[0] + b[1] * a[1] + b[2] * a[2] + b[3] * a[3]
            - b[4] * a[4]
            - b[5] * a[5]
            - b[6] * a[6]
            - b[7] * a[7];
        res[1] = b[1] * a[0] + b[0] * a[1] - b[4] * a[2] - b[5] * a[3] + b[2] * a[4] + b[3] * a[5]
            - b[7] * a[6]
            - b[6] * a[7];
        res[2] = b[2] * a[0] + b[4] * a[1] + b[0] * a[2] - b[6] * a[3] - b[1] * a[4]
            + b[7] * a[5]
            + b[3] * a[6]
            + b[5] * a[7];
        res[3] = b[3] * a[0] + b[5] * a[1] + b[6] * a[2] + b[0] * a[3]
            - b[7] * a[4]
            - b[1] * a[5]
            - b[2] * a[6]
            - b[4] * a[7];
        res[4] = b[4] * a[0] + b[7] * a[3] + b[0] * a[4] + b[3] * a[7];
        res[5] = b[5] * a[0] - b[7] * a[2] + b[0] * a[5] - b[2] * a[7];
        res[6] = b[6] * a[0] + b[7] * a[1] + b[0] * a[6] + b[1] * a[7];
        res[7] = b[7] * a[0] + b[0] * a[7];
        res
    }

    pub fn wedge(self: Self, b: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = b[0] * a[0];
        res[1] = b[1] * a[0] + b[0] * a[1];
        res[2] = b[2] * a[0] + b[0] * a[2];
        res[3] = b[3] * a[0] + b[0] * a[3];
        res[4] = b[4] * a[0] + b[2] * a[1] - b[1] * a[2] + b[0] * a[4];
        res[5] = b[5] * a[0] + b[3] * a[1] - b[1] * a[3] + b[0] * a[5];
        res[6] = b[6] * a[0] + b[3] * a[2] - b[2] * a[3] + b[0] * a[6];
        res[7] = b[7] * a[0] + b[6] * a[1] - b[5] * a[2] + b[4] * a[3] + b[3] * a[4] - b[2] * a[5]
            + b[1] * a[6]
            + b[0] * a[7];
        res
    }

    pub fn geometric_product(self: Self, b: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = b[0] * a[0] + b[1] * a[1] + b[2] * a[2] + b[3] * a[3]
            - b[4] * a[4]
            - b[5] * a[5]
            - b[6] * a[6]
            - b[7] * a[7];
        res[1] = b[1] * a[0] + b[0] * a[1] - b[4] * a[2] - b[5] * a[3] + b[2] * a[4] + b[3] * a[5]
            - b[7] * a[6]
            - b[6] * a[7];
        res[2] = b[2] * a[0] + b[4] * a[1] + b[0] * a[2] - b[6] * a[3] - b[1] * a[4]
            + b[7] * a[5]
            + b[3] * a[6]
            + b[5] * a[7];
        res[3] = b[3] * a[0] + b[5] * a[1] + b[6] * a[2] + b[0] * a[3]
            - b[7] * a[4]
            - b[1] * a[5]
            - b[2] * a[6]
            - b[4] * a[7];
        res[4] = b[4] * a[0] + b[2] * a[1] - b[1] * a[2] + b[7] * a[3] + b[0] * a[4] - b[6] * a[5]
            + b[5] * a[6]
            + b[3] * a[7];
        res[5] = b[5] * a[0] + b[3] * a[1] - b[7] * a[2] - b[1] * a[3] + b[6] * a[4] + b[0] * a[5]
            - b[4] * a[6]
            - b[2] * a[7];
        res[6] = b[6] * a[0] + b[7] * a[1] + b[3] * a[2] - b[2] * a[3] - b[5] * a[4]
            + b[4] * a[5]
            + b[0] * a[6]
            + b[1] * a[7];
        res[7] = b[7] * a[0] + b[6] * a[1] - b[5] * a[2] + b[4] * a[3] + b[3] * a[4] - b[2] * a[5]
            + b[1] * a[6]
            + b[0] * a[7];
        res
    }

    pub fn add(self: R300, b: R300) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0] + b[0];
        res[1] = a[1] + b[1];
        res[2] = a[2] + b[2];
        res[3] = a[3] + b[3];
        res[4] = a[4] + b[4];
        res[5] = a[5] + b[5];
        res[6] = a[6] + b[6];
        res[7] = a[7] + b[7];
        res
    }

    pub fn sub(self: R300, b: R300) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0] - b[0];
        res[1] = a[1] - b[1];
        res[2] = a[2] - b[2];
        res[3] = a[3] - b[3];
        res[4] = a[4] - b[4];
        res[5] = a[5] - b[5];
        res[6] = a[6] - b[6];
        res[7] = a[7] - b[7];
        res
    }

    pub fn divide(self: R300, b: R300) -> R300 {
        self.geometric_product(b.Reverse() * (1.0 / self.norm() * b.norm()))
    }
}

impl Not for R300 {
    type Output = R300;

    fn not(self: Self) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = -a[7];
        res[1] = -a[6];
        res[2] = a[5];
        res[3] = -a[4];
        res[4] = a[3];
        res[5] = -a[2];
        res[6] = a[1];
        res[7] = a[0];
        res
    }
}

// Mul
// The geometric product.
impl Mul for R300 {
    type Output = R300;

    fn mul(self: R300, b: R300) -> R300 {
        self.geometric_product(b)
    }
}

// Wedge
// The outer product. (MEET)
impl BitXor for R300 {
    type Output = R300;

    fn bitxor(self: R300, b: R300) -> R300 {
        self.wedge(b)
    }
}

// Dot
// The inner product.
impl BitOr for R300 {
    type Output = R300;

    fn bitor(self: R300, b: R300) -> R300 {
        self.dot(b)
    }
}

// Vee
// The regressive product. (JOIN)
// impl BitAnd for R300 {
//     type Output = R300;

//     fn bitand(self: R300, b: R300) -> R300 {
//         let mut res = R300::zero();
//         let a = self;
//         res[7] = 1 * (a[7] * b[7]);
//         res[6] = 1 * (a[6] * b[7] + a[7] * b[6]);
//         res[5] = -1 * (a[5] * -1 * b[7] + a[7] * b[5] * -1);
//         res[4] = 1 * (a[4] * b[7] + a[7] * b[4]);
//         res[3] = 1 * (a[3] * b[7] + a[5] * -1 * b[6] - a[6] * b[5] * -1 + a[7] * b[3]);
//         res[2] = -1 * (a[2] * -1 * b[7] + a[4] * b[6] - a[6] * b[4] + a[7] * b[2] * -1);
//         res[1] = 1 * (a[1] * b[7] + a[4] * b[5] * -1 - a[5] * -1 * b[4] + a[7] * b[1]);
//         res[0] = 1
//             * (a[0] * b[7] + a[1] * b[6] - a[2] * -1 * b[5] * -1 + a[3] * b[4] + a[4] * b[3]
//                 - a[5] * -1 * b[2] * -1
//                 + a[6] * b[1]
//                 + a[7] * b[0]);
//         res
//     }
// }

// Add
// Multivector addition
impl Add for R300 {
    type Output = R300;

    fn add(self: R300, b: R300) -> R300 {
        self.add(b)
    }
}

// Sub
// Multivector subtraction
impl Sub for R300 {
    type Output = R300;

    fn sub(self: R300, b: R300) -> R300 {
        self.sub(b)
    }
}

// smul
// scalar/multivector multiplication
impl Mul<R300> for f64 {
    type Output = R300;

    fn mul(self: f64, b: R300) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a * b[0];
        res[1] = a * b[1];
        res[2] = a * b[2];
        res[3] = a * b[3];
        res[4] = a * b[4];
        res[5] = a * b[5];
        res[6] = a * b[6];
        res[7] = a * b[7];
        res
    }
}

// muls
// multivector/scalar multiplication
impl Mul<f64> for R300 {
    type Output = R300;

    fn mul(self: R300, b: f64) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0] * b;
        res[1] = a[1] * b;
        res[2] = a[2] * b;
        res[3] = a[3] * b;
        res[4] = a[4] * b;
        res[5] = a[5] * b;
        res[6] = a[6] * b;
        res[7] = a[7] * b;
        res
    }
}

// sadd
// scalar/multivector addition
impl Add<R300> for f64 {
    type Output = R300;

    fn add(self: f64, b: R300) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a + b[0];
        res[1] = b[1];
        res[2] = b[2];
        res[3] = b[3];
        res[4] = b[4];
        res[5] = b[5];
        res[6] = b[6];
        res[7] = b[7];
        res
    }
}

// adds
// multivector/scalar addition
impl Add<f64> for R300 {
    type Output = R300;

    fn add(self: R300, b: f64) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0] + b;
        res[1] = a[1];
        res[2] = a[2];
        res[3] = a[3];
        res[4] = a[4];
        res[5] = a[5];
        res[6] = a[6];
        res[7] = a[7];
        res
    }
}

// ssub
// scalar/multivector subtraction
impl Sub<R300> for f64 {
    type Output = R300;

    fn sub(self: f64, b: R300) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a - b[0];
        res[1] = -b[1];
        res[2] = -b[2];
        res[3] = -b[3];
        res[4] = -b[4];
        res[5] = -b[5];
        res[6] = -b[6];
        res[7] = -b[7];
        res
    }
}

// subs
// multivector/scalar subtraction
impl Sub<f64> for R300 {
    type Output = R300;

    fn sub(self: R300, b: f64) -> R300 {
        let mut res = R300::zero();
        let a = self;
        res[0] = a[0] - b;
        res[1] = a[1];
        res[2] = a[2];
        res[3] = a[3];
        res[4] = a[4];
        res[5] = a[5];
        res[6] = a[6];
        res[7] = a[7];
        res
    }
}
