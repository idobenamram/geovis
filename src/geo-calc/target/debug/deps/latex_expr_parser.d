/Users/iamram/geovis/src/geo-calc/target/debug/deps/liblatex_expr_parser.dylib: /Users/iamram/latex-expr-parser/src/lib.rs /Users/iamram/latex-expr-parser/src/lexer.rs /Users/iamram/latex-expr-parser/src/parser.rs /Users/iamram/latex-expr-parser/src/token_set.rs

/Users/iamram/geovis/src/geo-calc/target/debug/deps/liblatex_expr_parser.rlib: /Users/iamram/latex-expr-parser/src/lib.rs /Users/iamram/latex-expr-parser/src/lexer.rs /Users/iamram/latex-expr-parser/src/parser.rs /Users/iamram/latex-expr-parser/src/token_set.rs

/Users/iamram/geovis/src/geo-calc/target/debug/deps/latex_expr_parser.d: /Users/iamram/latex-expr-parser/src/lib.rs /Users/iamram/latex-expr-parser/src/lexer.rs /Users/iamram/latex-expr-parser/src/parser.rs /Users/iamram/latex-expr-parser/src/token_set.rs

/Users/iamram/latex-expr-parser/src/lib.rs:
/Users/iamram/latex-expr-parser/src/lexer.rs:
/Users/iamram/latex-expr-parser/src/parser.rs:
/Users/iamram/latex-expr-parser/src/token_set.rs:
