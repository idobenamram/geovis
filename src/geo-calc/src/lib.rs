#![allow(unused_imports)]
#![allow(dead_code)]
#![allow(non_upper_case_globals)]
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]
pub mod r300;
use latex_expr_parser::{ASTNode, TokenKind};
use r300::R300;
use serde::Serialize;
use serde_wasm_bindgen::from_value;
use std::{collections::HashMap, ops::Mul};
use wasm_bindgen::prelude::*;


// TODO: this duplication is not my favorite
#[derive(Debug, Serialize)]
pub enum AstNodeWithValue {
    Identifier {
        name: String,
        value: R300,
    },
    Int {
        value: R300,
    },
    BinaryOpNode {
        op: TokenKind,
        left: Box<AstNodeWithValue>,
        right: Box<AstNodeWithValue>,
        value: R300,
    },
    UnaryOpNode {
        op: TokenKind,
        operand: Box<AstNodeWithValue>,
        value: R300,
    },
}

impl AstNodeWithValue {
    pub fn value(&self) -> R300 {
        match self {
            AstNodeWithValue::Int { value } => value.clone(),
            AstNodeWithValue::Identifier { value, .. } => value.clone(),
            AstNodeWithValue::BinaryOpNode { value, .. } => value.clone(),
            AstNodeWithValue::UnaryOpNode { value, .. } => value.clone(),
        }
    }
}

#[wasm_bindgen]
pub fn find_identifiers(expr: &str) -> Vec<String> {
    let ast = serde_json::from_str::<ASTNode>(expr).unwrap();
    let mut identifiers = Vec::new();
    find_ast_identifiers(&ast, &mut identifiers);
    identifiers
}

pub fn find_ast_identifiers(ast: &ASTNode, identifiers: &mut Vec<String>) {
    match ast {
        ASTNode::Identifier { name } => identifiers.push(name.clone()),
        ASTNode::BinaryOpNode { left, right, op: _ } => {
            find_ast_identifiers(left, identifiers);
            find_ast_identifiers(right, identifiers);
        }
        ASTNode::UnaryOpNode { operand, op: _ } => {
            find_ast_identifiers(operand, identifiers);
        }
        _ => {}
    }
}

#[wasm_bindgen]
pub fn calculate_expression(expr: &str, vars: JsValue) -> String {
    let ast = serde_json::from_str::<ASTNode>(expr).unwrap();
    let vars: HashMap<String, R300> = from_value(vars).unwrap();
    let calculated = calculate_ast_expression(&ast, &vars);
    serde_json::to_string(&calculated).unwrap()
}

fn calculate_ast_expression(ast: &ASTNode, vars: &HashMap<String, R300>) -> AstNodeWithValue {
    match ast {
        ASTNode::Int { value } => AstNodeWithValue::Int {
            value: R300::new(*value as f64, 0),
        },
        ASTNode::Identifier { name } => AstNodeWithValue::Identifier {
            name: name.clone(),
            value: vars[name].clone(),
        },
        ASTNode::BinaryOpNode { op, left, right } => {
            let left = calculate_ast_expression(left, vars);
            let right = calculate_ast_expression(right, vars);
            let left_val = left.value();
            let right_val = right.value();
            let value = match op {
                TokenKind::Plus => left_val.add(right_val),
                TokenKind::Minus => left_val.sub(right_val),
                TokenKind::Multiply => left_val.mul(right_val),
                TokenKind::Dot => left_val.dot(right_val),
                TokenKind::Wedge => left_val.wedge(right_val),
                TokenKind::Frac => left_val.divide(right_val),
                _ => panic!("Unsupported binary operator: {:?}", op),
            };
            AstNodeWithValue::BinaryOpNode {
                op: *op,
                left: Box::new(left),
                right: Box::new(right),
                value,
            }
        }
        ASTNode::UnaryOpNode { op, operand } => {
            let operand = calculate_ast_expression(operand, vars);
            let operand_val = operand.value();
            let value = match op {
                TokenKind::Plus => operand_val,
                TokenKind::Minus => -1.0 * operand_val,
                _ => panic!("Unsupported unary operator: {:?}", op),
            };
            AstNodeWithValue::UnaryOpNode {
                op: *op,
                operand: Box::new(operand),
                value,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_calculate_ast_expression() {
        let ast = ASTNode::BinaryOpNode {
            op: TokenKind::Plus,
            left: Box::new(ASTNode::Identifier {
                name: "a".to_string(),
            }),
            right: Box::new(ASTNode::Identifier {
                name: "b".to_string(),
            }),
        };
        let mut vars = HashMap::new();
        vars.insert("a".to_string(), R300::vector(1.0, 0.0, 0.0));
        vars.insert("b".to_string(), R300::vector(2.0, 0.0, 0.0));
        let result = calculate_ast_expression(&ast, &vars);

        insta::assert_debug_snapshot!(result);
    }
}
