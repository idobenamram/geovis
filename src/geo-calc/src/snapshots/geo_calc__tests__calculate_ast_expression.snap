---
source: src/lib.rs
expression: result
---
BinaryOpNode {
    op: Plus,
    left: Identifier {
        name: "a",
        value: R300 {
            mvec: [
                0.0,
                1.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
            ],
        },
    },
    right: Identifier {
        name: "b",
        value: R300 {
            mvec: [
                0.0,
                2.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
            ],
        },
    },
    value: R300 {
        mvec: [
            0.0,
            3.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
        ],
    },
}
